File: schema_vault/commands/up.py
import os
from pathlib import Path
from .base import get_connection, ensure_lock_table, get_applied, record_migration


def run(config: dict, dry_run: bool = False):
    conn = get_connection(config)
    ensure_lock_table(conn, config)

    applied = set(get_applied(conn, config))
    migrations_dir = Path(config.get("migrations_dir", "./migrations"))

    files = sorted(f for f in migrations_dir.glob("*.sql"))
    pending = [f for f in files if f.name not in applied]

    if not pending:
        print("Nothing to migrate - all up to date.")
        return

    for f in pending:
        content = f.read_text()
        parts = content.split("-- down")
        up_sql = parts[0].replace("-- up", "").strip()

        if dry_run:
            print(f"[dry-run] {f.name}:")
            print(up_sql)
            print()
            continue

        cur = conn.cursor()
        try:
            cur.execute(up_sql)
            conn.commit()
            record_migration(conn, config, f.name)
            print(f"Applied: {f.name}")
        except Exception as e:
            conn.rollback()
            print(f"Failed: {f.name} - {e}")
            raise
        finally:
            cur.close()

    conn.close()

File: src/commands/up.js
const fs = require('fs');
const path = require('path');
const { loadConfig } = require('../config');
const { getDriver } = require('../drivers');
const chalk = require('chalk');

exports.up = async (options) => {
  const config = loadConfig();
  const driver = getDriver(config);

  await driver.ensureLockTable();

  const applied = await driver.getAppliedMigrations();
  const appliedSet = new Set(applied.map(m => m.name));

  const files = fs.readdirSync(config.migrationsDir)
    .filter(f => f.endsWith('.sql'))
    .sort();

  const pending = files.filter(f => !appliedSet.has(f));

  if (!pending.length) {
    console.log(chalk.green('Nothing to migrate - all up to date.'));
    return;
  }

  for (const file of pending) {
    const content = fs.readFileSync(path.join(config.migrationsDir, file), 'utf8');
    const upMatch = content.match(/-- up([\s\S]*?)(?:-- down|$)/);
    if (!upMatch) continue;

    const sql = upMatch[1].trim();

    if (options.dryRun) {
      console.log(chalk.cyan(`[dry-run] ${file}:`));
      console.log(chalk.gray(sql));
      continue;
    }

    try {
      await driver.execute(sql);
      await driver.recordMigration(file);
      console.log(chalk.green(`Applied: ${file}`));
    } catch (err) {
      console.error(chalk.red(`Failed: ${file}`), err.message);
      process.exit(1);
    }
  }
};
